---
title: "RNAseq_pipeline_DESeq_Ebola_Vaccine_example"
author: "Richard Green"
date: "Friday, December 26, 2014"
output: html_document
RNASEQ pipeline
========================================================

Synopsis
--------
This is a RNAseq pipeline that will download data , run QC and generate statistics that can be used for Differential expression and Functional Analysis. This pipeline is based of the vignette for EdgeR and some online tutorials that can be found here:


http://cgrlucb.wikispaces.com/file/view/edgeR_Tutorial.pdf



First I will display my session info
```{r}
sessionInfo()
```

Download Data
-------------

Lets download the necessary library EDGER so we can analyze the data and then download the data.
The raw data was obtained from this study:

http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE56845



```{r}
#Lets download the library
#source("http://www.bioconductor.org/biocLite.R")
#install.packages("edgeR")
library(edgeR)
library(RCurl)
library(knitr)

#Lets download the file

raw.data <- read.csv(file="C:\\temp\\ebola_counts.txt", header = TRUE, sep="\t")


#we check to make sure our data loaded sucessfully
dim(raw.data)


#lets take a peek at the data set
head(raw.data)

#I am only interested in the Baseline samples and Infected samples at later time points
#I will select only those columns I am interested in.

counts <- raw.data[,c(3,7,11,5,9,13) ]


#I want to keep the gene names because I know I will need them later for analysis so I will store them
rownames( counts ) <- raw.data[ , 1 ] # gene names

#filter by an average of 10 counts or more

means <- rowMeans(counts)

filter <- means >= 10

table(filter)

counts <- counts[filter,]

dim(counts)



#I want to simply the column names to either baseline and Infected

colnames( counts ) <- paste(c(rep("Baseline",3),rep("Infected",3)),c(1:3,1:3),sep="") # sample names

#Lets see how the data has changed 
dim( counts )
```
Quality Control
---------------
We can look at the sum of each column and observe its library size. Next we can look for those genes that have very low gene counts because this will effect our analysis downstream. Once we see how many low expressed genes we have we will filter out those genes that have too few reads. We will keep genes that have at least 1 read per million in at least 3 samples

```{r}
colSums( counts ) # Library Sizes

colSums( counts ) / 1e06 # Library Sizes in millions of reads

table( rowSums( counts ) )[ 1:30 ] # Number of genes with low counts
head( counts )
group <- c(rep("Baseline", 3) , rep("Infected", 3))
cds <- DGEList( counts , group = group )
names( cds )
head(cds$counts) # original count matrix
cds$samples # contains a summary of your samples
sum( cds$all.zeros ) # How many genes have 0 counts across all samples
```
Normalization
-------------
We calculate normalization factors and effective libraries sizes. Normalization factors correct for the di???erences within the samples. Normalization factors are then multipled by the e???ective library size and the product becomes the library size which are displayed in the table below.

Below we apply normalization factors which correct for the di???erences within the samples.This is multipled with the e???ective library size are then the product of the actual library sizes and these factors.

```{r}

cds <- cds[rowSums(1e+06 * cds$counts/expandAsMatrix(cds$samples$lib.size, dim(cds)) > 1) >= 3, ]
dim( cds )
cds <- calcNormFactors( cds )
cds$samples

# effective library sizes
library_sizes <- cds$samples$lib.size * cds$samples$norm.factors

head(library_sizes)

```
Dimensionality Reduction
------------------------

Lets make a MDS plot(like a PCA) this will take all my now normalized counts and plot them in geometric space to look at relationship between the conditions, how well my replicates bind together, and how much variability exists between my samples and conditions.

```{r ,fig.width=10, fig.height=10,echo=TRUE}

plotMDS( cds , main = "MDS Plot for Count Data", labels = colnames( cds$counts ) )
```



```{r, echo=FALSE}
cds <- estimateCommonDisp( cds )
names( cds )
# The estimate
cds$common.dispersion

sqrt( 200 ) # poisson sd
sqrt( 200 + 200^2 * cds$common.dispersion ) # negative binomial sd
sqrt( 200 + 200^2 * cds$common.dispersion ) / sqrt( 200 ) #

cds <- estimateTagwiseDisp( cds , prior.df = 10 )
names( cds )
summary( cds$tagwise.dispersion )


# More shrinkage/sqeezing toward the common
cds <- estimateTagwiseDisp( cds , prior.df = 25 )

summary( cds$tagwise.dispersion ) # not much changed, but the ends got squeezed in quite a bit.

# The recommended setting for this data set is the default of 10. Let???s stick with that.
cds <- estimateTagwiseDisp( cds , prior.df = 10 )
```

Mean Variance Plot
------------------


```{r ,fig.width=10, fig.height=10,echo=TRUE}
plotMeanVar( cds , show.raw.vars=TRUE ,
show.tagwise.vars=TRUE ,
show.binned.common.disp.vars=FALSE ,
show.ave.raw.vars=FALSE ,
NBline = TRUE ,
nbins = 100 ,
pch = 16 ,
xlab ="Mean Expression (Log10 Scale)" ,
ylab = "Variance (Log10 Scale)" ,
main = "Mean-Variance Plot" )
```

Differential Expression
-----------------------

```{r}
de.cmn <- exactTest( cds , dispersion="common" , pair = c("Baseline" , "Infected" ) )
de.tgw <- exactTest( cds , dispersion="common" , pair = c("Baseline" , "Infected" ) )
de.poi <- exactTest( cds , dispersion = 1e-06 , pair = c("Baseline" , "Infected" ) )
names( de.tgw )
de.tgw$comparison # which groups have been compared
head( de.tgw$table ) # results table in order of your count matrix.
head( cds$counts )

options( digits = 3 ) # print only 3 digits
topTags( de.tgw , n = 20 , sort.by = "p.value" ) # top 20 DE genes
# Back to count matrix for tagwise analysis
cds$counts[ rownames( topTags( de.tgw , n = 15 )$table ) , ]

# Sort tagwise results by Fold-Change instead of p-value
resultsByFC.tgw <- topTags( de.tgw , n = nrow( de.tgw$table ) , sort.by = "logFC" )$table
head( resultsByFC.tgw )
# Store full topTags results table
resultsTbl.cmn <- topTags( de.cmn , n = nrow( de.cmn$table ) )$table
resultsTbl.tgw <- topTags( de.tgw , n = nrow( de.tgw$table ) )$table
resultsTbl.poi <- topTags( de.poi , n = nrow( de.poi$table ) )$table

# Names/IDs of DE genes
de.genes.cmn <- rownames( resultsTbl.cmn )[ resultsTbl.cmn$adj.P.Val <= 0.05 ]
de.genes.tgw <- rownames( resultsTbl.tgw )[ resultsTbl.tgw$adj.P.Val <= 0.05 ]
de.genes.poi <- rownames( resultsTbl.poi )[ resultsTbl.poi$adj.P.Val <= 0.05 ]

summary( decideTestsDGE( de.tgw , p.value = 0.05 ) )
```


Download my DE results here
---------------------------

```{r}

count_results_sorted <- cds$counts[ rownames( topTags( de.tgw , n = 30246 )$table ) , ]


colnames( resultsTbl.cmn ) <- c( "logConc" , "logFC" , "pVal.Cmn" , "adj.pVal.Cmn" )

colnames( resultsTbl.tgw ) <- c( "logConc" , "logFC" , "pVal.Tgw" , "adj.pVal.Tgw" )

#If we'd like, we could just write these tables to a file

#write.table(resultsTbl.cmn, file="${fileout:de_results}", sep = "\t",qmethod = "double", col.names=TRUE)


#write.table(resultsTbl.tgw, file = "${fileout:de_results}", sep = "\t",qmethod = "double", col.names=TRUE)

```
